### ISSUE TITLE
Container networking

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#4080]
## Overview

This issue tracks the design and implementation status of container networking, in its various forms.

| | Host-to-container (h2c) | Container-to-host (c2h) | Container-to-container (c2c) |
| -- | -- | -- | -- |
| **Use cases** | Expose a service to the host | <ul><li>SSH agent forwarding</li><li>Connect to docker engine</li><li>Connect to a service only reachable from the host</li></ul> | <ul><li>Run a test database</li><li>Run e2e integration tests</li><li>Run sidecar services</li></ul> |
| **Status** | ❌ not implemented | 🏗 in development | :heavy_check_mark: #4505 |
| **Notes** | Will only be available for long-running containerized services, not short-lived *exec* | <ul><li>Unix sockets: ✅ implemented</li><li>TCP: 🏗 API discussion</li><li>UDP: 🏗 API discussion</li></ul> | Use bass c2c networking as a reference |

## Design proposals

Container networking is a vast feature area. There are several ongoing proposals to design certain parts of it. Some of these proposals are compatible, while others are not.

| Proposal | c2h (tcp/udp) | h2c (tcp/udp) | c2c | c2h (unix) | h2c (unix) | status
| -- | -- | -- | -- | -- | -- | -- |
| [Mount unix socket into exec](https://github.com/dagger/dagger/issues/3850#issuecomment-1329899415) | ❌  | ❌ | ❌ | ✅  | ❌ | Implemented |
| [forwarding host TCP/UDP addresses to container](https://github.com/dagger/dagger/pull/4078) | ✅ | ❌ | ❌ | ❌ | ❌  |In Discussion |
| ~~[inbound and outbound tunnels](https://github.com/dagger/dagger/pull/4158)~~ | ~~✅~~ | ~~✅~~| ~~❌~~ | ~~❌~~ | ~~❌~~ | ~~Withdrawn~~ |
| [bass-style c2c networking](https://github.com/dagger/dagger/issues/4080#issuecomment-1344849016) | ❌ | ❌ | ✅ | ❌ | ❌ | Implemented |
| [“bass-plus” container networking](https://github.com/dagger/dagger/pull/4163) | ✅ | ✅ | ✅| ❌ | ❌| Partially implemented |

### Submitting a proposal

1. Open a draft PR with graphql schema changes and, ideally, skeleton documentation
2. Post a link to the PR in the comments to this issue. Mention which areas are covered by the proposal (c2h, h2c, c2c)

### Discussing a proposal

1. Comment on the relevant PR

### Research material

* SSH port forwarding (`ssh -L` and `ssh -R`)
* Reverse proxy configurations (nginx, apache, varnish, envoy)
* Kubernetes and Docker networking terminology
* Bass-lang networking

**Comment by sipsma** [comment]
Are we considering the container-to-host case (i.e. I want to run a dev webserver in dagger and access it from my browser) to be the same and host-to-container? In theory they are technically different (who does the connect, who does the listen), though if you do a reverse proxy thing locally then the can become the same.

---

Also, should probably clarify whether "container" means "exec" or "service" or both. My starting point is that we should not support serving sockets from execs and only serve them from services; execs are only able to consume them. But there's a lot to unpack and think about there so I'm far from adamant, just my first inclination.

**Comment by aluzzardi** [comment]
> My starting point is that we should not support serving sockets from execs and only serve them from services;

I agree -- had some comments about that in https://github.com/dagger/dagger/pull/3691#issuecomment-1311145961

**IF** that's the case, then we don't necessarily need to mirror all calls for bind/connect.

My straw man proposal for container-to-service was to mirror host:

```go
c.Container().WithTCPSocket(8080, c.Host().TCPSocket(8080))
c.Container().WithTCPSocket(8080, myService.TCPSocket(8080))
```

Regarding host-to-service (**IF** it's always a service), then we can have some service specific instructions:

```go
// 1) At service creation time
c.Service().Create(<args from the future>, dagger.ServiceCreateOpts{
  Expose: [some port bindings]
})

// 2) Dynamic exposure
// ??? Unsure how it'd play out ??
// Probably a bit too dynamic (once the service is running it's too late to mess with the shim, without some serious effort)
myService.TCPSocket(8080).ExposeTCP("localhost:8080")

// 3) CLI command
$ dagger service proxy <my service>:8080:8080
```

**Comment by jlongtine** [comment]
> Regarding host-to-service (**IF** it's always a service), then we can have some service specific instructions:

I like option 1, personally. Keeps it in code, keeps it at service "boot". The only suggestion I have is that we clarify that it is being exposed on the host, not just within the Buildkit network/context. I'm not quite sure how to do that yet, but just `Expose` feels a bit unclear to my eyes.

Now that I've written that down, I'm wondering if it's possible have an API like this:

```graphqls
type Host {
  bind|expose(service: Service!): ??
}
```

That would make pure GraphQL service queries a bit harder/uglier, but I like the clarity that would provide for the SDKs.

**Comment by vito** [comment]
+1 to starting with a distinction between services and containers for the sake of conversation.

Depending on how this goes, it _might_ be the case that the service's container needs to be aware that it's being used as a service - via the underlying LLB we generate and/or the responsibilities of the shim - but we can cross that bridge when we get to it.

I put a bunch of details around how Bass's c2c/service networking works here: https://github.com/dagger/dagger/issues/3850#issuecomment-1329899415

Here are some questions based on that:

**How do we want the traffic to flow in the following snippet?**

```go
c.Container().WithTCPSocket(8080, myService.TCPSocket(8080))
```

Typically a client container will need to be given the address to reach as a flag/arg/env/etc. With this scheme would they just use `127.0.0.1:8080`?

Does that imply we have something proxying traffic between them? Are we OK with the performance hit there? The answer might simply be yes, and maybe it's a minor performance hit anyhow, but it's worth trying to think about any performance-critical use cases here. If we only care about dev/test this is probably fine. It's hard to imagine someone running e.g. a production database with Dagger since it's not really built for persistence.

Another option is to use direct c2c addressing with a bridge network (+ DNS to prevent ephemeral container IPs from busting the cache) but then we'd need to figure out how the user configures the address for the client container. We can also use DNS to avoid cache busting while _keeping_ the proxy to solve that.

**Will we implement health/liveness/readiness checks?**

My take on this is the minimum viable health check is to just wait for a port to be listening, and that we should just do that by default since it's really annoying to have to manually poll for simple cases.

I don't have a strong opinion on continuous health monitoring beyond that.

---

@jlongtine Good question - I think we likely won't ever need to call 'expose' for container-to-container (within Buildkit network). So I think we can get away with having `Expose` live on the service and have it always mean "expose on the host" which would draw parallels to e.g. Container.Export, Directory.Export which also affect the the host.

**Comment by shykes** [comment]
I'm unable to discuss these examples constructively, because `WithTCPSocket` does not clarify whether I'm connecting or binding.

**Comment by jlongtine** [comment]
> @jlongtine Good question - I think we likely won't ever need to call 'expose' for container-to-container (within Buildkit network). So I think we can get away with having `Expose` live on the service and have it always mean "expose on the host" which would draw parallels to e.g. Container.Export, Directory.Export which also affect the the host.

Yeah, this makes sense, @vito. I'm on board.

**Comment by shykes** [comment]
Yesterday we had a 1-1 chat with @vito that melted my brain. After sleeping on it, I am updating this issue with:

1) A rewrite of the context and summary table. Hopefully it's more useful now
2) A list of open questions
3) A proposal for terminology and API design. To quote @vito: "i think there's potentially something better still, but it's a good strawman"

Please take a look and share your thoughts.

**Comment by shykes** [comment]
Addressing a few comments and questions above:

> @sipsma:
> Are we considering the container-to-host case (i.e. I want to run a dev webserver in dagger and access it from my browser) to be the same as host-to-container? In theory they are technically different (who does the connect, who does the listen), though if you do a reverse proxy thing locally then they can become the same.

👍 You are right, they are different and my framing was awkward. The root cause of the awkwardness was the use of the term "socket" which does actually apply to both (a socket is something you either connect or bind. then bytes are forwarded in both directions). Removing the term "socket" from feature name and API design allowed me to reconcile 1) non-confusing terminology for the thing being forwarded, and 2) non-confusing terminology for the "direction" of things.

> @sipsma:
> Also, should probably clarify whether "container" means "exec" or "service" or both. My starting point is that we should not support serving sockets from execs and only serve them from services; execs are only able to consume them. But there's a lot to unpack and think about there so I'm far from adamant, just my first inclination.
> @aluzzardi :
> I agree

👍 I agree also, clarified this in the issue.

For high-level discussion, it's still useful to use c2h/h2c/c2c (3 terms), instead of c2h/s2h/h2s/c2s/s2s (5 terms) :) But that's just terminology, we're all on the same page on design and implementation.

> @aluzzardi @vito @jlongtine 
> (on the concept of "expose" for h2c)

🙋🏻‍♂️It makes sense in theory, since we have "directory export" and "container export" which follow the same pattern.
However, I think we should *not* use this particular pattern here, because it is not intuitive whether the port is actually bound in the host, or simply available for ulterior service discovery from either the host or a container.

In Docker we had terminology for both: you could "expose" a port (available for service discovery), or "publish" it (bound to the host). This was generally confusing to users.

I propose a variation of this pattern that still controls the tunneling from `Service` (rather than `Host`), but is more explicit about what's going on. `withInboundTunnel` clearly means "create a tunnel that will forward inbound traffic".

**Comment by grouville** [comment]
With the implementation of socket forwarding, users have started to spawn local registries to their host docker engine, using them as a long-runner engine: https://github.com/dagger/dagger/issues/4089

By relying on Docker for most long-running tasks:
- the `Host-to-container (h2c)` is not as urgent as `Container-to-host (c2h)`. Same for the `Container-to-container (c2c)`, which could be workarounded by users if we had access to `c2h` primitives
- healthchecks are quite easy to implement natively and **might alternatively be tackled using extensions**:
```go
// query localhost:5000/v2/ until it returns 200 with a 20 seconds timeout
func waitForRegistry() error {
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, "GET", "http://localhost:5000/v2/", nil)
	if err != nil {
		return err
	}

	client := http.Client{}
	for {
		resp, err := client.Do(req)
		if err != nil {
			// retry
			time.Sleep(1 * time.Second)
			continue
		}

		if resp.StatusCode == http.StatusOK {
			return nil
		}
	}
}
```

**Current blockers**
I am currently relying on the `WithMountedSockets` API for #4038.

However, I am blocked by the fact that I cannot export an image (using our `core` API) to my private registry if it is not part of the `dagger-engine` network. I currently have 2 options: 
- create a registry that is part of the `dagger-engine` network
- push my image using the docker CLI

I feel that the `socket` implementation tackles well the SSH forwarding issues, but the `Container-to-host` begins to feel needed for more complex use-cases

**Comment by aluzzardi** [comment]
> Typically a client container will need to be given the address to reach as a flag/arg/env/etc. With this scheme would they just use 127.0.0.1:8080?
> Does that imply we have something proxying traffic between them? Are we OK with the performance hit there?

I remember reading the design in bass around services and it made a lot of sense. Do you have something in mind for this API that could avoid the proxying? We're now free to customize buildkit however we want

> Will we implement health/liveness/readiness checks?
> My take on this is the minimum viable health check is to just wait for a port to be listening, and that we should just do that by default since it's really annoying to have to manually poll for simple cases.

I think it would make sense to have a minimum viable health check. Otherwise, without the user implementing *something*, it's likely to fail every single time (e.g. create service mysql, exec query => instant failure).

This *might* require configuration -- we should at least make sure there's a "spot" in the API where that config could eventually fit.

> API design for container-to-host

```graphql
  withOutboundTunnel(protocol: TunnelProtocol=TCP, backend: String!, frontend: String): Container!
```

nit: This is inconsistent with the pattern of going through Host(), e.g. `WithDirectory("/foo", c.Host().Directory("."))`.

(not rhetorical) so why not `c.WithHostDirectory()` and getting rid of `c.Host().Directory()`?

Another factor is container-to-service. The reason I found `dagger.Socket` compelling is it gave a layer of indirection (disregard the word "socket") and a nice symmetry with directories:

```go
// FS
ctr.WithDirectory("/foo", dg.Host().Directory(".")) // c-to-h FS
ctr.WithDirectory("/bar", anotherCtr.Directory("/bar")) // c-to-c FS

// Network
ctr.WithSomeNetwork(something, dg.Host().SomeNetwork(8080)) // c-to-h NET
ctr.WithSomeNetwork(something, aService.SomeNetwork(8080)) // c-to-s NET
```

**Comment by vito** [comment]
> I remember reading the design in bass around services and it made a lot of sense. Do you have something in mind for this API that could avoid the proxying? We're now free to customize buildkit however we want

Here's a mockup, I think this as close as I can get it to Bass's setup:

```go
func TestServices(t *testing.T) {
	c, ctx := connect(t)
	defer c.Close()

	www := c.Directory().WithNewFile("index.html", "Hello, world!")

	// NB: for now I'm just making this a Container, ignoring the Container vs
	// Service argument to stay focused
	srv := c.Container().
		From("python").
		WithMountedDirectory("/srv/www", www).
		WithWorkdir("/srv/www").
		WithExec([]string{"python", "-m", "http.server"}).
		// optional: not a fundamental requirement, but I suspect named ports will
		// be easier to work with + refactor across abstraction boundaries. it also
		// tells Dagger what ports to health check.
		WithExposedPort("http", 8080)

	client := c.Container().
		From("alpine").
		// similar to WithSecretVariable. introduces a dependency between
		// client<->server and configures its address as an env var.
		WithServiceVariable("URL", srv.Address("http", "http://$host:$port")).
		WithExec([]string{"sh", "-c", "wget -O- $URL"})

	// running the client will:
	// 1. start the server as a buildkit solve(). we use a solve() so that
	//    concurrent runs are deduped. buildkit automatically does this + proxies
	//    progress to all clients.
	// 2. poll for the service's exposed ports to be bound.
	// 3. render the $URL ServiceVariable using the server's DNS address and port.
	//    note that this API would also support tunneling if we need to; just
	//    replace $host with '127.0.0.1' and $port with whatever ephemeral port
	//    we use in the container.
	// 4. run the client with $URL
	// 5. detach the server (cancel() ctx). note that we detach instead of
	//    stopping; this way any concurrent instances will continue until all
	//    clients detach.
	out, err := client.Stdout(ctx)
	require.NoError(t, err)
	require.Equal(t, "Hello, world!", out)
}
```

This is roughly equivalent to the following Bass script:

```clojure
(defn http-server [index]
  (from (linux/python)
    (-> ($ python -m http.server)
        (with-mount (mkfile ./index.html index) ./index.html)
        (with-port :http 8000))))

(defn main []
  (let [srv (http-server "Hello, world!")]
    (-> ($ wget -O- (addr srv :http "http://$host:$port"))
        (with-image (linux/alpine))
        (read :raw))))
```

Some notes:

* We have to use an env var and `sh -c` to pass a service address as a CLI arg, since we can't intermingle a complex address type with the regular string args.
* DNS addresses would _not_ be user-configurable, to avoid collisions. Instead they would be a content-addressed hash. For Bass I use base32 + xxh3 to get short addresses like `GRLO5DSIPLJPQ`.
* This involves the use of a template string for `$host` and `$port`, which is a little interesting. Open to other ideas there.
* This is all only concerned with IP addressing. I'm not sure to what extent we care about Unix sockets here.

**Comment by shykes** [comment]
@vito in your c2c design, are there constraints in how we implement IP addressing and routing? For example can the IP address be different between runs of otherwise identical services? Could the operator eventually configure IP ranges to use, or allocate them themselves via a hook, or would that break the design.

**Comment by shykes** [comment]
> nit: This is inconsistent with the pattern of going through Host(), e.g. `WithDirectory("/foo", c.Host().Directory("."))`.

But it's consistent with `Container.Export` and `Directory.Export` :)

> (not rhetorical) so why not `c.WithHostDirectory()` and getting rid of `c.Host().Directory()`?

I don't have a good answer. Empirically, this part of the current API has been fine in practice. I think it's because `Host().Directory()` is a great starting point for many different kinds of chains. Whereas a `Socket` type or equivalent isn't really chainable: there's only one place you can use it. So maybe the difference is in the level of composition we need?

**Comment by shykes** [comment]
> @vito in your c2c design, are there constraints in how we implement IP addressing and routing, and implementation of backing network interfaces? For example can the IP address be different between runs of otherwise identical services? Could the operator eventually configure IP ranges to use, or allocate them themselves via a hook, or would that break the design.

Answering myself with details from [this other comment](https://github.com/dagger/dagger/issues/3850#issuecomment-1329899415) : 

> Bass containers reach services by addressing them over a bridge network, using DNS to resolve to container hostnames to IPs at runtime.
> 
> DNS was my solution for not busting the cache. The key point is to avoid having ephemeral container IPs ever enter into LLB. I got DNS to work by configuring Buildkit to use CNI and installing dnsmasq and the dnsname plugin:

**Comment by sipsma** [comment]
> DNS addresses would not be user-configurable, to avoid collisions. Instead they would be a content-addressed hash. For Bass I use base32 + xxh3 to get short addresses like GRLO5DSIPLJPQ.

@vito What's the input to the hash you use in bass? The LLB of the service being depended on? If so, then I'm wondering how we'd deal with sources where the LLB is stable but the actual contents can change (e.g. `image:latest` or local dirs)

**Comment by vito** [comment]
> @vito What's the input to the hash you use in bass? The LLB of the service being depended on? If so, then I'm wondering how we'd deal with sources where the LLB is stable but the actual contents can change (e.g. `image:latest` or local dirs)

Bass uses the hash of the thunk, so it's a hash of the LLB "recipe" rather than the actual LLB. Thunks are fully specified - they pin image digests and don't use any other Git/HTTP LLB sources. Just images and commands.

So this is a great point. I think it would be fine to use a hash of the LLB even with unresolved references since they should be stable within that session, _as long as_ the single session is used for running both the service and the client. This definitely relates to #3613 especially if that's not actually possible for some reason (similar to the current `Container.Export` limitation).

**Comment by yonidavidson** [comment]
Adding my +1 to this feature, I would like to use this for running pre migration validations on databases and I need to spin up a MySQL DB and another container that communicates with it

**Comment by sourcec0de** [comment]
I've got a few use cases for this.
1. as @yonidavidson said, we have a pipeline that needs to run validations and migrations against a database
2. we need to access IPs that are only accessible from the host networking

Both locally and in Github actions we are able to reach these systems through tailscale.

**Comment by mircubed** [comment]
@vito with #4505 merged, can you update the issue description? It looks like it wasn't updated since Dec 13th, so I want to make sure that it reflects the work that still needs to be done. Thanks!
